### PR TITLE
Update tests (PHP 8 & PHPUnit 9)

### PR DESCRIPTION
**Pull request by iMattPro**

### CODE REVIEW

File: tests/analyser/run_test.php [hunk 1]
```diff
@@ -117,7 +117,7 @@ public function test_run($ad_code, $is_https, $expected)
 			$i = 0;
 			foreach ($expected as $message)
 			{
-				$this->template->expects($this->at($i))
+				$this->template->expects(self::at($i))
```
Comment (rxu) [inline tests/analyser/run_test.php:120]: `at()` matcher is deprecated in PHPUnit 9 and will be removed in v.10, so it'd be a good idea to replace it with something else (f.e. `exactly` matcher with related argument sequences etc).

File: tests/controller/helper_test.php [hunk 2]
```diff
@@ -293,13 +293,13 @@ public function test_assign_locations($ad_locations)
 				),
 			));
 
-		$this->template->expects($this->at(0))
+		$this->template->expects(self::at(0))
```
Comment (rxu) [inline tests/controller/helper_test.php:296]: Same as above about `at()` matcher.

File: tests/controller/helper_test.php [hunk 3]
```diff
@@ -293,13 +293,13 @@ public function test_assign_locations($ad_locations)
...
 			->method('assign_block_vars')
 			->with('ad_locations', array(
 				'CATEGORY_NAME'  => 'CAT_TOP_OF_PAGE',
 			));
 
-		$this->template->expects($this->at(1))
+		$this->template->expects(self::at(1))
```
Comment (rxu) [inline tests/controller/helper_test.php:302]: Same as above about `at()` matcher.

File: tests/controller/helper_test.php [hunk 4]
```diff
@@ -308,13 +308,13 @@ public function test_assign_locations($ad_locations)
 				'S_SELECTED'    => $ad_locations ? in_array('top_of_page_1', $ad_locations) : false,
 			));
 
-		$this->template->expects($this->at(2))
+		$this->template->expects(self::at(2))
```
Comment (rxu) [inline tests/controller/helper_test.php:311]: Same as above about `at()` matcher.

File: tests/controller/helper_test.php [hunk 5]
```diff
@@ -308,13 +308,13 @@ public function test_assign_locations($ad_locations)
...
 			->method('assign_block_vars')
 			->with('ad_locations', array(
 				'CATEGORY_NAME'  => 'CAT_BOTTOM_OF_PAGE',
 			));
 
-		$this->template->expects($this->at(3))
+		$this->template->expects(self::at(3))
```
Comment (rxu) [inline tests/controller/helper_test.php:317]: Same as above about `at()` matcher.

File: tests/event/views_test.php [hunk 6]
```diff
@@ -40,26 +40,26 @@ public function test_views_with_bots($is_bot)
...
-		$this->controller_helper->expects(($is_bot ? $this->never() : $this->once()))
+		$this->controller_helper->expects(($is_bot ? self::never() : self::once()))
 			->method('route')
 			->with('phpbb_ads_view', array('data' => '1'))
 			->willReturn('app.php/adsview/1');
 
 		if (!$is_bot)
 		{
-			$this->template->expects($this->at(1))
+			$this->template->expects(self::at(1))
```
Comment (rxu) [inline tests/event/views_test.php:62]: Same as above about `at()` matcher.

File: tests/functional/content_only_test.php [hunk 7]
```diff
@@ -41,22 +41,22 @@ public function test_content_only_ad_hides()
...
 		$crawler = self::request('GET', 'memberlist.php');
 		$this->assertContainsLang('MEMBERLIST', $crawler->filter('h2')->eq(1)->text());
-		$this->assertNotContains($ad_code, $crawler->html());
+		self::assertNotContains($ad_code, $crawler->html());
 
 		$crawler = self::request('GET', 'viewonline.php');
-		$this->assertCount(1, $crawler->filter('.viewonline-title'));
-		$this->assertNotContains($ad_code, $crawler->html());
+		self::assertCount(1, $crawler->filter('.viewonline-title'));
+		self::assertNotContains($ad_code, $crawler->html());
```
Comment (rxu) [inline tests/functional/content_only_test.php:60]: `assertStringNotContainsString()`?

File: tests/functional/end_date_test.php [hunk 8]
```diff
@@ -59,6 +59,6 @@ public function test_past_end_date_is_not_displayed()
 		$crawler = self::request('GET', 'index.php');
 
 		// Confirm below header ad is not present
-		$this->assertNotContains($ad_code, $crawler->html());
+		self::assertNotContains($ad_code, $crawler->html());
```
Comment (rxu) [inline tests/functional/end_date_test.php:62]: `assertStringNotContainsString()`?

File: tests/functional/locations_test.php [hunk 9]
```diff
@@ -42,15 +42,15 @@ public function test_location_after_first_post()
 		$crawler = self::request('GET', 'viewtopic.php?t=1');
 
 		// Confirm after first post ad is NOT after first post when it's the only post
-		$this->assertNotContains($ad_code, $crawler->filter('#p1')->nextAll()->eq(0)->html());
+		self::assertNotContains($ad_code, $crawler->filter('#p1')->nextAll()->eq(0)->html());
```
Comment (rxu) [inline tests/functional/locations_test.php:45]: `assertStringNotContainsString()`?

File: tests/functional/start_date_test.php [hunk 10]
```diff
@@ -59,6 +59,6 @@ public function test_future_start_date_is_not_displayed()
 		$crawler = self::request('GET', 'index.php');
 
 		// Confirm below header ad is not present
-		$this->assertNotContains($ad_code, $crawler->html());
+		self::assertNotContains($ad_code, $crawler->html());
```
Comment (rxu) [inline tests/functional/start_date_test.php:62]: `assertStringNotContainsString()`?

**Review by rxu — COMMENTED** [review]
Sorry for the long review :P

File: tests/analyser/run_test.php (hunk 1 above)
Comment (rxu) [reply tests/analyser/run_test.php:120]: I was struggling with something like this in the core tests here, could be probably useful as an idea, starting from this line https://github.com/phpbb/phpbb/pull/6022/commits/f1c178b5447bbea380faf794514823916d0b961a#diff-fb115cbd979b76d90c4b772ae9c902a4bb4bb078c2f21aee7868784f320d7d1eR68 till the line 82.

Comment (iMattPro) [reply tests/analyser/run_test.php:120]: Can you send me that as a PR? I was truggling too, didn't even realize the ... thing for the array as arguments

Comment (rxu) [reply tests/analyser/run_test.php:120]: > Can you send me that as a PR?

@iMattPro Sure, will do it this weekend.

**Comment by rxu** [comment]
> Error: Access to undeclared static property PHPUnit\Framework\Error\Warning::$enabled

`\tests\test_framework\phpbb_test_case_helpers.php` -> `function setExpectedTriggerError()` needs to be fixed. Core tests pass as long as they don't test against`E_NOTICE`/`E_STRICT`/`E_WARNING`.

**Comment by rxu** [comment]
Just a guess, `Undefined array key 0` error is probably because the modules data isn't in the database in `MODULES_TABLE`. I wonder if it will give any benefit from making the test to extend `phpbb_database_test_case` and apply fixture on the `MODULES_TABLE` with modules data.

**Comment by rxu** [comment]
The rest 3 failures like `Failed asserting that exception message 'Undefined array key "id"' contains ...` backtraces are:
```
phpbb\ads\controller\admin_controller_test::test_action_add_submit with data set #3 (false, 2)
Undefined array key "id"
/home/travis/build/phpBB3/phpBB/includes/acp/auth.php:855
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/controller/admin_controller.php:533
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/controller/admin_controller.php:457
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/controller/admin_controller.php:182
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/controller/admin_controller.php:162
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/tests/controller/admin_controller_test.php:591
```
```
phpbb\ads\controller\admin_controller_test::test_action_edit_submit with data set #2 (false, true, 2)
Undefined array key "id"
/home/travis/build/phpBB3/phpBB/includes/acp/auth.php:855
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/controller/admin_controller.php:533
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/controller/admin_controller.php:482
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/controller/admin_controller.php:214
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/controller/admin_controller.php:162
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/tests/controller/admin_controller_test.php:912
```
```
phpbb\ads\controller\admin_controller_test::test_action_delete with data set #3 (1, 2, false, true)
Undefined array key "id"
/home/travis/build/phpBB3/phpBB/includes/acp/auth.php:855
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/controller/admin_controller.php:533
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/controller/admin_controller.php:279
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/controller/admin_controller.php:162
/home/travis/build/phpBB3/phpBB/ext/phpbb/ads/tests/controller/admin_controller_test.php:1042
```

The problem is that `$this->acl_options['id']` is not defined in the tests and it can't be straightly fixed as `$this->auth_admin` in admin controller is protected property. Probably using PHP Reflection in tests would help.

**Pull request merged**